specs/lazy-adr/adr-010-exec-api.md

```diff
@@ -0,0 +1,265 @@
+
+```go
+// Executor defines a common interface for interacting with the execution client.
+type Executor interface {
+	// InitChain initializes the blockchain with genesis information.
+	InitChain(ctx context.Context, genesisTime time.Time, initialHeight uint64, chainID string) (stateRoot types.Hash, maxBytes uint64, err error)
+
+	// GetTxs retrieves all available transactions from the execution client's mempool.
+	GetTxs(ctx context.Context) ([]types.Tx, error)
+
+	// ExecuteTxs executes a set of transactions to produce a new block header.
+	ExecuteTxs(ctx context.Context, txs []types.Tx, blockHeight uint64, timestamp time.Time, prevStateRoot types.Hash) (updatedStateRoot types.Hash, maxBytes uint64, err error)
+
+	// SetFinal marks a block at the given height as final.
+	SetFinal(ctx context.Context, blockHeight uint64) error
+}
+```
+
+### API Methods specification
+
+#### `InitChain`
+
+##### Description
+
+Initializes the blockchain's state based on genesis information. This method is invoked at the beginning of the blockchain's lifecycle to prepare the execution environment for subsequent operations.
+
+##### Inputs
+
+- `ctx` (`context.Context`): Context for managing request timeouts and cancellations.
+- `genesisTime` (`time.Time`): The initial timestamp of the rollup.
+- `initialHeight` (`uint64`): The starting height of the rollup.
+- `chainID` (`string`): A unique identifier of the rollup network.
+
+##### Outputs
+
+- `stateRoot` (`types.Hash`): The resulting state root after initializing the chain.
+- `maxBytes` (`uint64`): Maximum block size in bytes, as defined by the genesis configuration.
+- `err` (`error`): An error, if the initialization process fails.
+
+##### Expected Behavior
+
+- Initialize rollup according to the genesis.
+- Generate an initial `stateRoot` representing the genesis state of the rollup.
+- Return the maximum allowable block size (`maxBytes`).
+
+#### `GetTxs`
+
+##### Description
+
+Fetches all pending transactions from the execution client's mempool.
+Transactions returned by execution client will be passed by rollkit to sequencer.
+
+##### Inputs
+
+- `ctx` (`context.Context`): Context for managing request timeouts and cancellations.
+
+##### Outputs
+
+- `txs` (`[]types.Tx`): Slice of transactions retrieved from the execution client's mempool; ordering doesn't matter.
+- `err` (`error`): An error, if any, while retrieving transactions.
+
+##### Expected Behavior
+
+- Access the mempool and retrieve all available transactions.
+- If no transactions are available, return an empty slice without error.
+- Do not remove ("reap") transactions from mempool.
+
+#### `ExecuteTxs`
+
+##### Description
+
+Executes a given set of transactions, updating the blockchain state.
+
+##### Inputs
+
+- `ctx` (`context.Context`): Context for managing request timeouts and cancellations.
+- `txs` (`[]types.Tx`): A slice of transactions to be executed.
```

Comment:
what is the type here?

Reply:
`types.Tx` is just `[]byte`, as it's not interpreted by rollkit in any way. I'll include information about types in ADR.

Reply:
why not use []byte? what is the type preventing here?

Reply:
TBH there is no good reason for this. ADR was reflecting code. We should change this in docs and code. Especially after https://github.com/rollkit/go-sequencing/pull/35